src/controllers/hid/hidcontroller.cpp

```diff
@@ -243,13 +244,27 @@ int HidController::close() {
 bool HidController::poll() {
     Trace hidRead("HidController poll");
 
-    int result = hid_read(m_pHidDevice, m_pPollData, sizeof(m_pPollData) / sizeof(m_pPollData[0]));
-    if (result == -1) {
-        return false;
-    } else if (result > 0) {
-        Trace process("HidController process packet");
-        QByteArray outData(reinterpret_cast<char*>(m_pPollData), result);
-        receive(outData, mixxx::Time::elapsed());
+    int result = 1;
+    while (result > 0) {
+        result = hid_read(m_pHidDevice, m_pPollData, sizeof(m_pPollData) / sizeof(m_pPollData[0]));
+        if (result == -1) {
+            return false;
+        } else if (result > 0) {
+            Trace process("HidController process packet");
+            auto byteArray = QByteArray::fromRawData(reinterpret_cast<char*>(m_pPollData), result);
+            // Some controllers such as the Gemini GMX continuously send input packets even if it
+            // is identical to the previous packet. If this loop processed all those redundant
+            // packets, it would be a big performance problem.
+            // TODO: Test if this hack can be done with adequate performance in JS so a deep copy
+            // of each incoming packet isn't needed for most HID devices.
+            if (byteArray == m_lastIncomingData) {
+                continue;
+            } else {
+                // force a copy
+                m_lastIncomingData = QByteArray(byteArray.data());
+            }
+            receive(byteArray, mixxx::Time::elapsed());
```

Comment:
This invocation will always cause a deep copy because the passed QByteArray doesn't own the data and receive() has the wrong signature, i.e. the QByteArray is passed by (const) value instead of by const reference.

Suggestion: Fix the signature of Controller::receive()

Reply:
Not your fault. The intention was good, but the legacy code defeats it :laughing:

Reply:
Sorry for the confusion, seems to work as intended. Implicit sharing also works even if the data is not owned. This is very dangerous. But, well, it's C++ not Rust.

Reply:
But if receive() decides to store a copy that lives longer than this invocation it would crash. So, not sane at all, independent of how the QByteArray parameter is passed.

I was hoping that Qt prevented shallow copying of a raw QByteArray and would perform a deep copy instead. But without the restriction the function call is not valid. Otherwise you need to document in the base class that any implementation ***must not*** copy and store the first parameter for longer than the function call.

Reply:
Nice catch. So we should not use an explicit shared container in this case. 
Would it be possible to just pass the raw pointer or a std::vector() or such.

Reply:
This would require refactoring that IMO should be beyond the scope of this urgent bugfix for the beta branch. We can come back to optimizing this in master after #2920 is merged.

Reply:
Since the data behind the QByteArray is statically allocated, it is reasonably safe. The data in the array might change between receive() calls, that's the worst thing that could happen.